docs/development/best-practices.md

```diff
@@ -0,0 +1,154 @@
+- Avoid `[Array|Set|Map].prototype.forEach`. It makes code harder to debug and defeats some useful compiler checks like reachability
+
+Only use `Array.prototype.map()` when the return value is used, otherwise use `for ( ... of ...)`.
+
+[Source](https://google.github.io/styleguide/tsguide.html#iterating-objects)
+
+### Exports
+
+Use named exports in all code.
+Avoid default `exports`.
+This way all `imports` follow the same pattern.
+
+[Source, reasoning and examples.](https://google.github.io/styleguide/tsguide.html#exports)
+
+### Imports
+
+Use [ES6 module](https://exploringjs.com/es6/ch_modules.html#sec_basics-of-es6-modules) syntax, i.e.
+
+```ts
+import { square, diag } from 'lib';
+
+// You may also use:
+
+import * as lib from 'lib';
+```
+
+And avoid `require`:
+
+```ts
+import x = require('...');
+```
+
+### HTTP & RESTful API request handling
+
+Prefer using `Http` from `util/http` to simplify HTTP request handling and to enable authentication and caching, As our `Http` class will transparently handle host rules.
+Example:
+
+```ts
+import { Http } from '../../../util/http';
+
+const http = new Http('some alias');
+
+try {
+    const body = (await http.getJson<Response>(url)).body;
+} catch (err) {
+  ...
+}
+```
+
+### Dates and times
+
+Use [`Luxon`](https://www.npmjs.com/package/luxon) to handle dates and times.
+Use `UTC` to be time zone independent.
+
+[Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/distro.ts#L133-L134)
+
+### Unit testing
+
+- Use `it.each` rather than `test.each`
+- Prefer [Tagged Template Literal](https://jestjs.io/docs/api#2-testeachtablename-fn-timeout) style for `it.each`, Prettier will help with formatting
+  - See [Example](https://github.com/renovatebot/renovate/blob/768e178419437a98f5ce4996bafd23f169e530b4/lib/modules/platform/util.spec.ts#L8-L18)
+- Mock Date/Time when testing a Date/Time dependent module
+  - For `Luxon` mocking see [Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/distro.spec.ts#L7-L10)
+- Prefer `jest.spyOn` for mocking single functions, or mock entire modules
+  - Avoid overwriting functions, for example: (`func = jest.fn();`)
+
+### Fixtures
+
+- Use `Fixture` class for loading fixtures
+
+```ts
+Fixture.get('./file.json'); // for loading raw data
+Fixture.getJson('./file.json'); // for loading and parsing objects
+```
+
+### Working with vanilla JS files (renovate/tools only)
+
+Use [JSDoc](https://jsdoc.app/index.html) to declare types and function prototypes.
+
+[Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/tools/distro-json-generate.mjs#L7-L17)
+
+### Classes
+
+Use [Typescript getter setters (Accessors) when needed](https://google.github.io/styleguide/tsguide.html#properties-used-outside-of-class-lexical-scope).
+The getter must be a `pure function`.
```

Comment:
what is `pure` 🤔

Reply:
should we add an explanation ? i.e

pure functions
 - are predictable: given the same input, always returns the same output.
 - have no side effect

Reply:
Please add that explanation. 😉

Reply:
I've added the explanation given in the provided wiki page. 
hope its alright